L1Trigger/TrackerTFP/src/DataFormats.cc

```diff
@@ -0,0 +1,707 @@
+    get<6>(data_) = this->phiT() * setup->mhtNumBinsPhiT() + phiT;
+    get<7>(data_) = this->qOverPt() * setup->mhtNumBinsQoverPt() + qOverPt;
+    get<1>(data_) += base(Variable::qOverPt) * (qOverPt - .5) * r() - base(Variable::phiT) * (phiT - .5);
+    dataFormats_->convertStub(data_, frame_.second, p_);
+    fillTrackId();
+  }
+
+  void StubMHT::fillTrackId() {
+    TTBV ttBV(bv());
+    trackId_ = ttBV.extract(width(Variable::sectorPhi) + width(Variable::sectorEta) + width(Variable::phiT) + width(Variable::qOverPt));
+  }
+
+  StubSF::StubSF(const TTDTC::Frame& frame, const DataFormats* formats) :
+    Stub(frame, formats, Process::sf), chi2_(0.)
+  {
+    fillTrackId();
+  }
+
+  StubSF::StubSF(const StubMHT& stub, int layer, int zT, int cot, double chi2) :
+    Stub(stub, stub.r(), stub.phi(), 0., layer, stub.sectorPhi(), stub.sectorEta(), stub.phiT(), stub.qOverPt(), zT, cot), chi2_(chi2)
+  {
+    const double zTD = format(Variable::zT).floating(zT);
+    const double cotD = format(Variable::cot).floating(cot);
+    const Setup* setup = dataFormats_->setup();
+    get<2>(data_) = stub.z() - (zTD + cotD * (stub.r() + dataFormats_->chosenRofPhi() - setup->chosenRofZ()));
+    dataFormats_->convertStub(data_, frame_.second, p_);
+    fillTrackId();
+  }
+
+  void StubSF::fillTrackId() {
+    TTBV ttBV(bv());
+    trackId_ = ttBV.extract(width(Variable::sectorPhi) + width(Variable::sectorEta) + width(Variable::phiT) + width(Variable::qOverPt) + width(Variable::zT) + width(Variable::cot));
+  }
+
+  StubKFin::StubKFin(const TTDTC::Frame& frame, const DataFormats* formats, int layer) :
+    Stub(frame, formats, Process::kfin),
+    layer_(layer)
+  {
+    trackId_ = trackId();
+  }
+
+  StubKFin::StubKFin(const StubSF& stub, int layer, int trackId) :
+    Stub(stub, stub.r(), stub.phi(), stub.z(), trackId),
+    layer_(layer)
+  {
+    dataFormats_->convertStub(data_, frame_.second, p_);
+    trackId_ = trackId;
+  }
+
+  StubKFin::StubKFin(const TTStubRef& ttStubRef, const DataFormats* dataFormats, double r, double phi, double z, int trackId, int layer) :
+    Stub(ttStubRef, dataFormats, Process::kfin, r, phi, z, trackId),
+    layer_(layer)
+  {
+    dataFormats_->convertStub(data_, frame_.second, p_);
+    trackId_ = trackId;
+  }
+
+  StubKF::StubKF(const TTDTC::Frame& frame, const DataFormats* formats, int layer) :
+    Stub(frame, formats, Process::kf), layer_(layer) {}
+
+  StubKF::StubKF(const StubKFin& stub, double qOverPt, double phiT, double cot, double zT) :
+    Stub(stub, stub.r(), stub.phi(), stub.z()),
+    layer_(stub.layer())
+  {
+    const Setup* setup = dataFormats_->setup();
+    get<1>(data_) -= phiT - this->r() * qOverPt;
+    get<2>(data_) -= zT + (this->r() + setup->chosenRofPhi() - setup->chosenRofZ()) * cot;
+    dataFormats_->convertStub(data_, frame_.second, p_);
+  }
+
+  template<typename ...Ts>
+  Track<Ts...>::Track(const FrameTrack& frame, const DataFormats* dataFormats, Process p) :
```

Comment:
The constructors of the Track class use extremely dense C++. Please add a comment to each one explaining what it is used for, and what it is doing.

Reply:
done

Reply:
Better, but it remains unclear what "variables"  means in the comments "ntuple of variables".  As the argument of ntuple is a templated type, it is not immediately obvious from the C++ what the comment means by "variable".  e.g. If it is a floating point number or a digitized one, what it represents etc.

Reply:
Could you please give me suggestion how I shall change the comment?

Reply:
DataFormats.h & .cc use many times "typename ... Ts" or ntuple<Ts...>. The reader has no idea (without spending a lot of time investigating the code) what type of template argument "Ts" will be used here. e.g. Will it be an "int" or a "float" or one of your classes? Or what it physically represents? Without this info, it's extremely hard to understand what this code is doing.
I believe ntuple contains all the variables of a Stub or Track etc., in non-digitized form, and that "Ts" is a mix of int and float. If so, this needs stating somewhere clearly.